src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Repository/LocationRepository.cs

```diff
@@ -0,0 +1,315 @@
+                         WHERE {cols.Select(col => $"{col} != ''").JoinToString(" AND "):raw}
+                         GROUP BY {cols.JoinToString(", "):raw}
+                         ORDER BY {cols.JoinToString(", "):raw}
+                         """
+                    ).QueryAsync(cancellationToken: cancellationToken)
+                )
+                .Cast<IDictionary<string, object?>>()
+                .Select(row => row.ToDictionary(
+                    kv => kv.Key,
+                    kv => (string)kv.Value!
+                ))
+                .Select(row => MapLocationOptionMeta(row, meta.Level).ToRow())
+                .ToList();
+
+            var optionTable = publicDataDbContext
+                .GetTable<LocationOptionMetaRow>()
+                .TableName(nameof(PublicDataDbContext.LocationOptionMetas));
+
+            var current = 0;
+
+            const int batchSize = 1000;
+
+            while (current < options.Count)
+            {
+                var batch = options
+                    .Skip(current)
+                    .Take(batchSize)
+                    .ToList();
+
+                // We create a 'row key' for each option that allows us to quickly
+                // find the option rows that exist in the database. It's typically
+                // much slower to have multiple WHERE clauses for each row that check
+                // against every other row. Out of several such attempts, the 'row key'
+                // technique was the fastest and simplest way to create the links.
+                var batchRowKeys = batch
+                    .Select(o => o.GetRowKey())
+                    .ToHashSet();
+
+                Expression<Func<LocationOptionMetaRow, bool>> hasBatchRowKey =
+                    o => o.Type == batch[0].Type &&
+                         batchRowKeys.Contains(
+                             o.Type + ',' +
+                             o.Label + ',' +
+                             (o.Code ?? "null") + ',' +
+                             (o.OldCode ?? "null") + ',' +
+                             (o.Urn ?? "null") + ',' +
+                             (o.LaEstab ?? "null") + ',' +
+                             (o.Ukprn ?? "null")
+                         );
+
+                var existingRowKeys = (await optionTable
+                        .Where(hasBatchRowKey)
+                        .ToListAsync(token: cancellationToken))
+                    .Select(o => o.GetRowKey())
+                    .ToHashSet();
+
+                if (existingRowKeys.Count != batch.Count)
+                {
+                    var newOptions = batch
+                        .Where(o => !existingRowKeys.Contains(o.GetRowKey()))
+                        .ToList();
+
+                    var startIndex = await publicDataDbContext.LocationOptionMetas.CountAsync(token: cancellationToken);
```

Comment:
Similar thing to earlier comment - this relies on not running multiple instances of this function in parallel.

We'll need to consider how to prevent parallel runs, or come up with some alternative solution to prevent duplicates and gaps in the integer sequence.